File: apps/worldai/src/sdk/python/worldai_client.py
import asyncio
import httpx
import websockets
import json
from typing import Callable, Optional, Dict, Any, Awaitable

class WorldAIClient:
    def __init__(self, base_url: str = "http://localhost:8000"):
        self.base_url = base_url.rstrip('/')
        self.ws_url = self.base_url.replace("http://", "ws://").replace("https://", "wss://")
        self.client = httpx.AsyncClient(base_url=self.base_url)
        self._ws_connection: Optional[Any] = None
        self._ws_task: Optional[asyncio.Task] = None

    async def get_world(self) -> Dict[str, Any]:
        """현재 세계 상태 기본 정보를 가져옵니다."""
        response = await self.client.get("/world")
        response.raise_for_status()
        return response.json()

    async def get_factions(self) -> list:
        """전체 파벌 정보를 가져옵니다."""
        response = await self.client.get("/factions")
        response.raise_for_status()
        return response.json()

    async def get_leaderboard(self) -> Dict[str, Any]:
        """군사력, 영향력 등 파벌 리더보드를 반환합니다."""
        response = await self.client.get("/leaderboard")
        response.raise_for_status()
        return response.json()

    async def tick(self, hours: int = 1) -> Dict[str, Any]:
        """지정된 시간(시간 단위)만큼 시뮬레이션을 진행시킵니다."""
        response = await self.client.post("/simulation/tick", params={"hours": hours})
        response.raise_for_status()
        return response.json()

    async def run_simulation(self, hours: int, active: bool = True) -> Dict[str, Any]:
        """지정된 시간만큼 백그라운드 처리를 실행하거나 중지합니다."""
        response = await self.client.post("/simulation/run", json={"hours": hours, "active": active})
        response.raise_for_status()
        return response.json()

    async def connect_events(self, on_message: Callable[[Dict[str, Any]], Awaitable[None]]):
        """WebSocket에 연결하여 실시간 이벤트(영토 변경, 이벤트 로그 등) 수신을 비동기로 시작합니다."""
        if self._ws_task is not None and not self._ws_task.done():
            return  # Already connected

        async def _listen():
            try:
                async with websockets.connect(f"{self.ws_url}/ws") as websocket:
                    self._ws_connection = websocket
                    async for message in websocket:
                        data = json.loads(message)
                        # Call the user-provided callback
                        await on_message(data)
            except Exception as e:
                print(f"[WorldAIClient] WebSocket connection closed or error: {e}")
            finally:
                self._ws_connection = None

        self._ws_task = asyncio.create_task(_listen())

    async def close(self):
        """클라이언트 및 소켓 연결을 종료합니다."""
        if self._ws_connection is not None:
            await self._ws_connection.close()
        if self._ws_task is not None:
            self._ws_task.cancel()
        await self.client.aclose()

File: apps/worldai/src/sdk/csharp/WorldAIClient.cs
using System;
using System.Net.Http;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Tasks;
using System.Collections.Concurrent;
using System.Collections.Generic;

namespace WorldAI.SDK
{
    public class WorldAIClient : IDisposable
    {
        private readonly HttpClient _httpClient;
        private readonly Uri _wsUri;
        private ClientWebSocket _webSocket;
        private CancellationTokenSource _cts;

        // Thread-safe message queue for Unity main-thread consumption
        private readonly ConcurrentQueue<string> _messageQueue;

        public WorldAIClient(string baseUrl = "http://localhost:8000")
        {
            _httpClient = new HttpClient { BaseAddress = new Uri(baseUrl) };
            
            var wsUrl = baseUrl.Replace("http://", "ws://").Replace("https://", "wss://").TrimEnd('/');
            _wsUri = new Uri($"{wsUrl}/ws");
            _messageQueue = new ConcurrentQueue<string>();
        }

        /// <summary>
        /// Gets the current state of the world.
        /// </summary>
        public async Task<string> GetWorldAsync()
        {
            var response = await _httpClient.GetAsync("/world");
            response.EnsureSuccessStatusCode();
            return await response.Content.ReadAsStringAsync();
        }

        /// <summary>
        /// Gets the data for all factions.
        /// </summary>
        public async Task<string> GetFactionsAsync()
        {
            var response = await _httpClient.GetAsync("/factions");
            response.EnsureSuccessStatusCode();
            return await response.Content.ReadAsStringAsync();
        }

        /// <summary>
        /// Advances the simulation by specified hours.
        /// </summary>
        public async Task<string> TickAsync(int hours = 1)
        {
            var response = await _httpClient.PostAsync($"/simulation/tick?hours={hours}", null);
            response.EnsureSuccessStatusCode();
            return await response.Content.ReadAsStringAsync();
        }

        /// <summary>
        /// Connects to the WebSocket and starts receiving real-time event updates into the concurrent queue.
        /// </summary>
        public async Task ConnectEventsAsync()
        {
            if (_webSocket != null && _webSocket.State == WebSocketState.Open)
                return;

            _webSocket = new ClientWebSocket();
            _cts = new CancellationTokenSource();

            await _webSocket.ConnectAsync(_wsUri, _cts.Token);
            _ = ReceiveMessagesAsync(); // Fire and forget the listening loop
        }

        private async Task ReceiveMessagesAsync()
        {
            var buffer = new byte[1024 * 16]; // 16KB buffer for delta updates
            try
            {
                while (_webSocket.State == WebSocketState.Open && !_cts.IsCancellationRequested)
                {
                    var result = await _webSocket.ReceiveAsync(new ArraySegment<byte>(buffer), _cts.Token);
                    if (result.MessageType == WebSocketMessageType.Close)
                    {
                        await _webSocket.CloseAsync(WebSocketCloseStatus.NormalClosure, "Closing", _cts.Token);
                    }
                    else if (result.MessageType == WebSocketMessageType.Text)
                    {
                        string message = Encoding.UTF8.GetString(buffer, 0, result.Count);
                        _messageQueue.Enqueue(message);
                    }
                }
            }
            catch (OperationCanceledException)
            {
                // Normal closure
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[WorldAIClient] WebSocket Error: {ex.Message}");
            }
        }

        /// <summary>
        /// Consumes all received event messages from the background queue.
        /// Recommended to be called from Unity's Update() loop to process UI safely on the main thread.
        /// </summary>
        public IEnumerable<string> ConsumeEvents()
        {
            while (_messageQueue.TryDequeue(out var message))
            {
                yield return message;
            }
        }

        public void DisconnectEvents()
        {
            _cts?.Cancel();
            _webSocket?.Abort();
            _webSocket?.Dispose();
            _webSocket = null;
        }

        public void Dispose()
        {
            DisconnectEvents();
            _httpClient?.Dispose();
        }
    }
}
